src/rdmnet/broker/broker_client.cpp

```diff
@@ -389,3 +448,78 @@ bool RPTDevice::Send(const etcpal::Uuid& broker_cid)
   }
   return false;
 }
+
+void RPTDevice::ClearAllQueues()
+{
+  rpt_msgs_.clear();
+  status_msgs_.clear();
+  broker_msgs_.clear();
+}
+
+bool RPTDevice::RptMsgQ::empty() const
+{
+  return total_msg_count_ == 0;
+}
+
+MessageRef* RPTDevice::RptMsgQ::front()
+{
+  // Fair scheduler - we iterate through the controller map in order, starting from the last
+  // controller serviced.
+  auto con_pair = rpt_msgs_.upper_bound(current_controller_);
+  if (con_pair == rpt_msgs_.end())
+    con_pair = rpt_msgs_.begin();
+  while (con_pair->first != current_controller_)
+  {
+    if (!con_pair->second.empty())
+    {
+      // Found a controller ready to service.
+      break;
+    }
+
+    if (++con_pair == rpt_msgs_.end())
+      con_pair = rpt_msgs_.begin();
+  }
+  if (!con_pair->second.empty())
+  {
+    current_controller_ = con_pair->first;
+    return &con_pair->second.front();
+  }
+  return nullptr;
+}
+
+void RPTDevice::RptMsgQ::pop_front()
+{
+  if (current_controller_ != kInvalidHandle)
+  {
+    rpt_msgs_[current_controller_].pop_front();
+    --total_msg_count_;
+  }
+}
+
+void RPTDevice::RptMsgQ::push_back(Handle controller, MessageRef&& value)
+{
+  rpt_msgs_[controller].push_back(std::move(value));
+  ++total_msg_count_;
+}
+
+size_t RPTDevice::RptMsgQ::size() const
+{
+  return total_msg_count_;
+}
+
+void RPTDevice::RptMsgQ::RemoveCurrentController()
+{
+  auto controller_pair = rpt_msgs_.find(current_controller_);
+  if (controller_pair != rpt_msgs_.end())
+  {
+    total_msg_count_ -= controller_pair->second.size();
+    rpt_msgs_.erase(controller_pair);
+  }
```

Comment:
Should we reset current_controller_ here?

Reply:
I don't think it's necessary - the fair scheduling logic uses `map::upper_bound()`, so the next pop will just select the next available controller.